std/experimental/typecons.d

```diff
@@ -1074,3 +1074,307 @@ pure nothrow @system unittest
+ * - static typeof(this) fromExpectedValue(T expectedValue)
+ * - static typeof(this) fromUnexpectedValue(E unexpectedValue)
+ *
+ * TODO swap
+ *
+ * TODO which functions should be `nothrow`?
+ *
+ * TODO later on: remove _ok when `_expectedValue` and ` _unexpectedValue` can store this state
+ * "collectively" for instance when both are pointers or classes (use trait
+ * `isAddress`)
+ *
+ * TODO ok for default construction to initialize
+ * - _expectedValue = T.init (zeros)
+ * - _ok = true (better to have _isError so default is zero bits here aswell?)
+ */
+struct Expected(T, E = Exception)
+if (!isInstanceOf!(Unexpected, T)) // an `Unexpected` cannot be `Expected` :)
+{
+    import std.algorithm.mutation : moveEmplace;
+
+    /// Construct from expected value `expectedValue.`
+    this(T expectedValue) @trusted
+    {
+        moveEmplace(expectedValue, _store.expectedValue);
+        _ok = true;
+    }
+    /// ditto
+    this(ref T expectedValue)
+    {
+        _store.expectedValue = expectedValue; // TODO do we need to care about `_store.expectedValue` init value??
+        _ok = true;
+    }
+
+    /// Construct from unexpected value `unexpectedValue.`
+    this(Unexpected!E unexpectedValue) @trusted
+    {
+        moveEmplace(unexpectedValue, _store.unexpectedValue);
+        _ok = false;
+    }
+    /// ditto
+    this(ref Unexpected!E unexpectedValue)
+    {
+        _store.unexpectedValue = unexpectedValue; // TODO do we need to care about `_store.unexpectedValue` init value??
+        _ok = false;
+    }
+
+    /// Assign from expected value `expectedValue.`
+    void opAssign()(auto ref T expectedValue) @trusted
+    {
+        // TODO is this ok?:
+        clear();
+        moveEmplace(expectedValue, _store.expectedValue);
+        _ok = true;
+    }
+
+    /// Assign from unexpected value `unexpectedValue.`
+    void opAssign(E unexpectedValue) @trusted
+    {
+        clear();
+        moveEmplace(unexpectedValue, _store.unexpectedValue);
+        _ok = false;
+    }
+
+    /// Clear (empty) contents.
+    private void clear() @trusted
+    {
+        release();
+        static if (isAddress!T)
+        {
+            _store.expectedValue = null;
+        }
+    }
+
+    /// Release any memory used to store contents.
+    private void release() @trusted
+    {
+        import std.traits : hasElaborateDestructor;
+        if (hasExpectedValue)
+        {
+            static if (!is(T == class))
+            {
+                static if (hasElaborateDestructor!T)
+                {
+                    destroy(_store.expectedValue);
+                }
+            }
+            _ok = false;
+        }
+        else
+        {
+            static if (!is(E == class))
+            {
+                static if (hasElaborateDestructor!E)
+                {
+                    destroy(_store.unexpectedValue);
+                }
+            }
+            destroy(_store.unexpectedValue);
```

Comment:
If this destroy call is correct, the one above it is unnecessary. If the one above is correct, this one is a bug. Either way, something is wrong here.